src/ast.hpp

```diff
@@ -225,6 +225,42 @@ struct TypeTerm : public AST {
 	    : AST {ASTTag::TypeTerm} {}
 };
 
+struct TypeVar : public AST {
+	Token const* m_token;
+
+	std::string const& text() {
+		return m_token->m_text;
+	}
+
+	TypeVar()
+	    : AST {ASTTag::TypeVar} {}
+};
+
+struct Union : public AST {
```

Comment:
This is not really an Union, but a union expression, or a union literal. Maybe change the name to reflect that?